extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java

```diff
@@ -0,0 +1,233 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.guacamole.auth.quickconnect.utility;
+
+import java.io.UnsupportedEncodingException;
+import java.lang.StringBuilder;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URLDecoder;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.apache.guacamole.GuacamoleClientException;
+import org.apache.guacamole.GuacamoleServerException;
+import org.apache.guacamole.GuacamoleException;
+import org.apache.guacamole.protocol.GuacamoleConfiguration;
+
+/**
+ * A utility class to parse out a URI into the settings necessary
+ * to generate a GuacamoleConfiguration object.
+ */
+public class QCParser {
+
+    /**
+     * The default protocol to use if one is not provided in
+     * the incoming URI.
+     */
+    public static final String DEFAULT_URI_PROTOCOL = "ssh";
+
+    /**
+     * The default host to use if one is not defined.
+     */
+    public static final String DEFAULT_URI_HOST = "localhost";
+
+    /**
+     * The regex to use to split username and password.
+     */
+    private static final Pattern userinfoPattern = Pattern.compile("(^[^:]+):?(.*)");
+
+    /**
+     * The regex group of the username.
+     */
+    private static final int USERNAME_GROUP = 1;
+
+    /**
+     * THe regex group of the password.
+     */
+    private static final int PASSWORD_GROUP = 2;
+
+    /**
+     * Parse out a URI string and get a GuacamoleConfiguration
+     * from that string, or an exception if the parsing fails.
+     *
+     * @param uri
+     *     The string form of the URI to be parsed.
+     *
+     * @return
+     *     A GuacamoleConfiguration using a combination of the parsed
+     *     URI values and default values when not specified in the
+     *     URI.
+     *
+     * @throws GuacamoleException
+     *     When an error occurs parsing the URI.
+     */
+    public static GuacamoleConfiguration getConfiguration(String uri)
+            throws GuacamoleException {
+
+        URI qcUri;
+        try {
+            qcUri = new URI(uri);
+        }
+        catch (URISyntaxException e) {
+            throw new GuacamoleClientException("Invalid URI Syntax", e);
+        }
+
+        // Break out individual components of the URI.
+        String protocol = qcUri.getScheme();
+        String host = qcUri.getHost();
+        int port = qcUri.getPort();
+        String userInfo = qcUri.getUserInfo();
+        String query = qcUri.getQuery();
+
+        String username = null;
+        String password = null;
+        Map<String, String> queryParams = null;
```

Comment:
There are a few cases here of the general form:

```java
String foo = null;

if (someCondition) {
    foo = somethingThatNeverReturnsNull();
}

if (foo != null) {
    doMoreThings(foo);
}
```

This is really taking what is effectively one condition and splitting it across two separate `if` blocks. The unnecessary initialization and unnecessary split can both be removed by collapsing into:

```java
if (someCondition) {
    String foo = somethingThatNeverReturnsNull();
    doMoreThings(foo);
}
```

The logic is a bit tangled otherwise.

Reply:
Okay, I'll take a look at this section and try to untangle it.

Reply:
Untangled.